protoarray/src/main/java/tech/pegasys/teku/protoarray/ProtoArray.java

```diff
@@ -212,14 +225,21 @@ public void maybePrune(Bytes32 finalizedRoot) {
     for (int nodeIndex = 0; nodeIndex < finalizedIndex; nodeIndex++) {
       Bytes32 root =
           checkNotNull(nodes.get(nodeIndex), "ProtoArray: Invalid node index").getBlockRoot();
-      indices.remove(root);
+      rootAndSlotIndices.remove(new BlockRootAndSlot(root, nodes.get(nodeIndex).getBlockSlot()));
+      rootIndices.remove(root);
     }
 
     // Drop all the nodes prior to finalization.
     nodes.subList(0, finalizedIndex).clear();
 
     // Adjust the indices map.
-    indices.replaceAll(
+    rootAndSlotIndices.replaceAll(
+        (key, value) -> {
+          int newIndex = value - finalizedIndex;
+          checkState(newIndex >= 0, "ProtoArray: New array index less than 0.");
+          return newIndex;
+        });
+    rootIndices.replaceAll(
```

Comment:
As discussed offline it makes me nervous that we have to process through both maps as pruning is reasonably expensive already.

@rolfyone talked it through and came up with the idea that we want a `Map(blockRoot -> List<Indices>)` structure (likely wrapped in a custom class to provide a nice interface).  The `List<Indices>` would be the ordered indices for each `ProtoNode` with this block root.  The first entry would be the slot the block was actually in and each subsequent entry would be the index for each subsequent empty slot.

Look up by block root then functions by simply taking the first index in the list.
Look up by block root and slot takes the index at position (slot - blockSlot).  The blockSlot is available from the `ProtoNode` from the first entry (likely should just be stored in the wrapper class for simple access).

With this structure we can do a block or block slot lookup in constant time and avoid creating a new instance of `SlotAndBlockRoot` for each lookup.

It also allows for the possibility in the future to switch to a sparse index for blocks that are followed by an extremely large number of empty slots - in that case you'd likely windup using a `NavigableMap<slot, index>` and not have to populate an index for all empty slots.